src/communication/base_communicator.hpp

```diff
@@ -0,0 +1,149 @@
+#pragma once
+
+#include <algorithm>
+#include <iostream>
+#include <vector>
+#include <random>
+#include <functional>
+
+#include <algorithms.hpp>
+#include <connection.hpp>
+#include <communication/gathered_vector.hpp>
+#include <event_queue.hpp>
+#include <spike.hpp>
+#include <util/debug.hpp>
+#include <util/double_buffer.hpp>
+#include <util/partition.hpp>
+#include <util/compare.hpp>
+
+namespace nest {
+namespace mc {
+namespace communication {
+
+// When the communicator is constructed the number of target groups and targets
+// is specified, along with a mapping between local cell id and local
+// target id.
+//
+// The user can add connections to an existing communicator object, where
+// each connection is between any global cell and any local target.
+//
+// Once all connections have been specified, the construct() method can be used
+// to build the data structures required for efficient spike communication and
+// event generation.
+template <typename CommunicationPolicy>
+class base_communicator {
+public:
+    using communication_policy_type = CommunicationPolicy;
+    using time_type = spike::time_type;
+    using id_type = cell_gid_type;
+
+    /// per-cell group lists of events to be delivered
+    using event_queue =
+        std::vector<postsynaptic_spike_event<time_type>>;
+
+    using gid_partition_type =
+        util::partition_range<std::vector<id_type>::const_iterator>;
+
+    struct spike_extractor {
```

Comment:
Yeah -- I had left that as an open question in the earlier revision. On the one hand, the less_than<extractor> is an extra step of templating. On the other, it really should be a generic lambda rather than a struct for extraction + the <. That's unnecessarily noisy, when all I really want to say is "apply < to id_type and spike.source". lt_spike_source, though, should be:
```
struct lt_spike_source {
    using id_type = typename spike::id_type;                                                                                                                                                                        
    bool operator()(const id_type& l, const spike& r) { return l<r.source; }
    bool operator()(const id_type& l, const spike& r) { return l.source<r; }
};
```

C++11 is still missing generic lambdas that could handle this.